rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,154 @@
+# KVDB/KVS Introspection APIs
+
+## Problem
+
+HSE does not currently have any way to introspect KVDB/KVS information.
+
+## Requirements
+
+HSE should provide APIs for the following:
+
+- KVDB
+  - create parameters
+  - runtime parameters
+  - home
+  - KVS names (done)
+- KVS
+  - create parameters
+  - runtime parameters
+  - name
+
+Any REST APIs will support the notion of multiple KVDBs.
+
+## Non-Requirements
+
+- Will not add support for multiple KVDBs in other areas of HSE
+
+## Solution
+
+Create new APIs in C, Python, and REST in order to expose this information. REST
+APIs are necessary because you cannot introspect a KVDB open in another process.
+
+### Overview
+
+#### C APIs
+
+The following C APIs will be added behind the experimental wall:
+
+```c
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the kvdb_home member in ikvdb struct.
+ */
+const char *
+hse_kvdb_home_get(const struct hse_kvdb *kvdb);
+
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the name member in ikvs struct.
+ */
+const char *
+hse_kvs_name_get(const struct hse_kvs *kvs);
+
+/* Returns a JSON representation of the parameter value, ie the value in
```

Comment:
No. User would need to pass "durability.enabled".

> I actually prefer this API to return a comma separated k/v pairs. Definitely easier to parse than JSON.

All this API does is return whatever the value in `key=value` is, which just happens to be valid JSON.

Reply:
The comment regarding returning a JSON representation threw me off. It's good that we can query only a single parameter using this API. Make sense now, thanks!

Reply:
No problem. The implementation of `key=value` parsing takes the pointer to `value` and calls `cJSON_Parse()` on it. It isn't extremely obvious to someone who doesn't know the implementation. The only parameter that really looks like JSON is mclass_policies, and since no one ever uses that, no one would really know.